lxd/container_lxc.go

```diff
@@ -3690,6 +3755,30 @@ func (c *containerLXC) cleanupNetworkDevices(netns string) {
 			continue
 		}
 
+		d, err := device.New("ct", m)
+		if err != nil && err != device.ErrUnsupportedDevType {
+			logger.Errorf("Failed to cleanup device %s: %v", k, err)
+			return
+		}
+
+		if err != device.ErrUnsupportedDevType {
+			volatile := c.deviceVolatileGet(k)
+			volatile["netns"] = netns // Pass in netns path as pseudo volatile.
+			saveVolatile, err := d.Stop(volatile)
```

Comment:
@stgraber I'm thinking that the Stop() and Start() functions could take a variable called `env` rather than `volatile`. It would still contain the device's volatile data, but we may also include other vars, like the `netns` in the line above. Using `env` would indicate that it isn't just volatile data.

Reply:
So for that part, I was thinking, would it be cleaner to have the device struct hold pointers to functions to retrieve and store state?

So when we instantiate a new Device, we'd pass it one function to retrieve all stored state and one function to update its recorded state. We can then have the container_lxc.go side of it feed functions that use the volatile storage.

That would save us from having to return those and have the caller handle the saving.

Reply:
@stgraber yes that makes sense, effectively we would be passing in the equivalent of the `deviceVolatileGet` function above. However in cases where we need to make modifications to the "environment" variables passed in, such as in the case above with `netns`, we could define a temporary callback function that calls the normal volatile retrieval function `deviceVolatileGet` and then adds the additional vars, before being passed into function.

For saving the returned data, my only note of concern around passing in a callback to the device is that this would allow the device to call it as many times as it wanted. This is both a good and bad thing, in that it allows the device more flexibility to store things more frequently, but I remember you saying that saving data to dqlite cluster was quite an expensive operation (relative speaking) and my intention with returning the data to be persisted was that it would put the responsibility (and control of how often it is persisted) back to LXD itself, at most frequently once per device start/stop.

Reply:
@stgraber one reason I made the volatile/env data passed in at start/stop time rather than device instantiation is to ensure it gets the latest copy of the data (as the device struct may have been instantiated earlier to validate config and be started later), and also to indicate that the volatile/env data is temporarily used just for starting/stopping and is not stored, unlike when the config is passed in.

Reply:
@tomponline The current volatile saving code is quite optimized as it doesn't cause a full call to Update as it used to be. So I'm not too worried about allowing calls to that from the device set code.

Passing functions to read/write volatile data also doesn't prevent the caller (container_lxc) to pass functions which allow for batching. I don't think we'll need that at this point, but if we wanted we could define two local wrappers in container_lxc for volatile read/set which then don't call the actual volatile set but instead update a map with us using a single deferred function to then update the database at the end of all device processing.

Reply:
Seperately passing an environment dict seems reasonable, though I guess we'll see exactly when that is needed and whether that's an indication that we didn't segment devices and container properly.

Reply:
@stgraber yes indeed, now that the detach/rename functionality is moved back into container_lxc.go (for now), there's no need to pass netns into the device interface.

Reply:
@stgraber add callback function for saving volatile data.